sgx-traits/src/lib.rs

```diff
@@ -0,0 +1,94 @@
+// Copyright 2019 Red Hat, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::marker::PhantomData;
+use std::io::Result;
+
+use sgx_types::{secs::Secs, tcs::Tcs, sigstruct::SigStruct};
+use sgx_types::secinfo::{SecInfo, Flags as Permissions};
+use bitflags::bitflags;
+
+/// An offset reference with neither read nor write capabilities
+///
+/// The Handle struct allows the creation of an opaque reference to a
+/// type that cannot be read or written. Neither the lifetime nor the
+/// type are discarded. This allows us to refer to an offset inside
+/// an enclave without fear that it will be dereferenced. The size of
+/// the Handle is always 64 bits with natural alignment. Therefore,
+/// the handle type can be embedded in structs.
+#[repr(transparent)]
+#[derive(Debug)]
+pub struct Handle<'h, T>(u64, PhantomData<&'h T>);
+
+impl<'h, T> Handle<'h, T> {
+    pub unsafe fn new(ptr: usize) -> Self {
+        Handle(ptr as u64, PhantomData)
+    }
+}
+
+bitflags! {
+    pub struct Flags: u64 {
+        const MEASURE = 1 << 0;
+    }
+}
+
+pub trait Enclave<'e> {
+    fn enter(&self, handle: &mut Handle<'e, Tcs>) -> Result<()>;
+}
+
+pub trait Builder<'b>: Sized {
+    type Enclave: Enclave<'b>;
+
+    fn new(secs: Secs) -> Result<Self>;
+
+    fn add_tcs(
+        &mut self,
+        tcs: Tcs,
+        offset: usize
+    ) -> Result<Handle<'b, Tcs>>;
+
+    fn add_struct<T>(
```

Comment:
Same question here about the difference between `add_struct` and `add_slice`.

Reply:
`add_slice()` is the kernel API. The other two are convenience wrapper functions.